openmicroscopy/openmicroscopy
# Modify group membership only

# What this PR does

Allows restricted Admins with only ```ModifyGroup``` privilege or only ```ModifyGroupMembership``` to use the edit-Group form without problems.
See https://trello.com/c/vX34H2wn/73-lightadmin-and-webadmin-ui

All the form functionality has been refactored, so needs a complete re-test:

# Testing this PR

1. Scenarios for user with only ```ModifyGroup``` privilege:
    - Create new group - Form should only show name, description & permissions (not add owners/members).
    - First omit the group name and Save, then enter the name of an existing group and Save (should see suitable error each time)
    - Then enter all valid details and Save. Should be returned to All Groups table with new Group created
    - Edit the group and try again to enter an existing group name or no group name and check for warnings.
    - Then enter valid new name and Save.
 1. [Now Login as root and create a new user who is ONLY in the above group]
 1. Scenarios for user with ```ModifyGroupMembership``` privilege:
    - Viewing webadmin/groups/ table, should see no "Create Group" button, but can edit groups.
    - Edit the group created above - Form should only show Owners and Members fields.
    - Add yourself and other users to be group members / owners & Save
    - Try to remove the user who is ONLY in this group - should see a suitable warning (and form should refresh to add user back).
    - Should be able to remove yourself from this group (but not from 'system' group).
 1. Scenarios for user with ```ModifyGroup``` AND ```ModifyGroupMembership``` privilege:
    - Create a new Group and add owners & members in one shot
    - Edit Group and Group Membership at the same time, testing Save with various invalid inputs as above.

![screen shot 2017-08-17 at 10 19 11](https://user-images.githubusercontent.com/900055/29405413-bdc596e4-8335-11e7-95da-ab2c6085aa2a.png)

![screen shot 2017-08-17 at 10 14 47](https://user-images.githubusercontent.com/900055/29405420-c301170a-8335-11e7-8400-5e6ef22f1d70.png)
